src/miniscript/evaluate.rs

```diff
@@ -0,0 +1,606 @@
+                match top{
+                    StackElement::Satisfied => Ok(Primitives{pks: vec![], hashlocks: vec![], timelocks: vec![]}),
+                    StackElement::Dissatisfied => sub.interpret(secp, stack, sighash, age),
+                    _ =>  Err(Error::CouldnotEvaluate)
+                }
+            },
+            AstElem::Unlikely(ref sub) => {
+                let top = stack.pop().expect("Witness stack has at least one element");
+                match top{
+                    StackElement::Dissatisfied => Ok(Primitives{pks: vec![], hashlocks: vec![], timelocks: vec![]}),
+                    StackElement::Satisfied => sub.interpret(secp, stack, sighash, age),
+                    _ =>  Err(Error::CouldnotEvaluate)
+                }
+            },
+            AstElem::AndCat(ref left, ref right) => {
+                let mut res = left.interpret(secp, stack, sighash, age)?;
+                res.join(&right.interpret(secp, stack, sighash, age)?);
+                Ok(res)
+            }
+            AstElem::AndBool(ref left, ref right) => interpret_and_bool(secp, stack, sighash, age, left, right),
+            AstElem::AndCasc(ref left, ref right) => interpret_and_casc(secp, stack, sighash, age, left, right),
+            AstElem::OrBool(ref left, ref right) => interpret_or_bool(secp, stack, sighash, age, left, right),
+            AstElem::OrCasc(ref left, ref right) |
+            AstElem::OrCont(ref left, ref right) => interpret_or_casc(secp, stack, sighash, age, left, right),
+            AstElem::OrKey(ref left, ref right) => interpret_or_key(secp, stack, sighash, age, left, right),
+            AstElem::OrKeyV(ref left, ref right) => {
+                let res = interpret_or_key(secp, stack, sighash, age, left, right)?;
+                evaluate_verify(stack)?;
+                Ok(res)
+            }
+            AstElem::OrIf(ref left, ref right) =>  interpret_or_if(secp, stack, sighash, age, left, right),
+            AstElem::OrIfV(ref left, ref right) => {
+                let res = interpret_or_if(secp, stack, sighash, age, left, right)?;
+                evaluate_verify(stack)?;
+                Ok(res)
+
+            }
+            AstElem::OrNotif(ref left, ref right) => interpret_or_if(secp, stack, sighash, age, right, left),
+            AstElem::Thresh(k, ref subs) => evaluate_thres(secp, stack, sighash,age, k, subs),
+            AstElem::ThreshV(k, ref subs) => {
+                let res = evaluate_thres(secp, stack, sighash,age, k, subs)?;
+                evaluate_verify(stack)?;
+                Ok(res)
+            }
+        }
+    }
+
+}
+
+/// This functions is equivalent to VERIFY. This pops the top element and checks if it is non-zero.
+/// Gives an error in all other cases
+fn evaluate_verify(stack: &mut Vec<StackElement>) -> Result<(), Error>{
+    let top = stack.pop().expect("Non-zero element for verify");
```

Comment:
These `expect`s all make me nervous. Suggest changing the whole function body to
```
stack.pop() == StackElement::Satisfied
```
(This is also a logic change, from "anything but `Dissatisfied`" to "must be `Satisfied`", which I think is more correct ... in particular there are some non-`Dissatisfied` pushes that the Script interpreter will evaluate as false, and in general I'd prefer to be as strict as we can possibly get away with.)

Reply:
I also think (though maybe you disagree) that it'd look a bit nicer to `impl` all these methods on `Vec<StackElement>`, like
```
impl Vec<StackElement> {
    fn verify(&mut self) -> Result<(), Error> {
        if stack.pop() == Some(StackElement::Satisfied) {
            Ok(())
        } else {
            Err(Error::CouldNotEvaluate)
        }
    }
}
```